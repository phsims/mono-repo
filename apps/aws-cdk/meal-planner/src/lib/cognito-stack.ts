import { Stack, StackProps, RemovalPolicy } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { UserPool, UserPoolClient } from 'aws-cdk-lib/aws-cognito';

export class CognitoStack extends Stack {
  public readonly userPool: UserPool;
  public readonly userPoolClient: UserPoolClient;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // Cognito Pool
    const userPool = new UserPool(this, 'userPool', {
      userPoolName: `${id}-pool`,
      removalPolicy: RemovalPolicy.DESTROY,
    });
    this.userPool = userPool;

    // App Client
    const userPoolClient = new UserPoolClient(this, 'userPoolClient', {
      userPool: userPool,
      userPoolClientName: `${id}-client`,
      authFlows: {
        userPassword: true,
      },
    });
    this.userPoolClient = userPoolClient;
  }
}
