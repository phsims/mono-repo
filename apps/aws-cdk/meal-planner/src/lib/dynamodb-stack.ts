import { Stack, RemovalPolicy } from 'aws-cdk-lib';
import {
  AttributeType,
  GlobalSecondaryIndexProps,
  Table,
  BillingMode,
} from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';

export class DynamoDbStack extends Stack {
  public readonly table: Table;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    //setup table
    const table = new Table(this, `${id}-table`, {
      partitionKey: { name: 'PK', type: AttributeType.STRING },
      billingMode: BillingMode.PAY_PER_REQUEST,
      removalPolicy: RemovalPolicy.DESTROY,
      sortKey: { name: 'SK', type: AttributeType.STRING },
    });
    this.table = table;

    // setup indexs
    const gsi1: GlobalSecondaryIndexProps = {
      indexName: 'GSI1',
      partitionKey: {
        name: 'GSI1PK',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'GSI1SK',
        type: AttributeType.STRING,
      },
    };
    table.addGlobalSecondaryIndex(gsi1);

    const gsi2: GlobalSecondaryIndexProps = {
      indexName: 'GSI2',
      partitionKey: {
        name: 'GSI2PK',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'GSI2SK',
        type: AttributeType.NUMBER,
      },
    };

    table.addGlobalSecondaryIndex(gsi2);
  }
}
