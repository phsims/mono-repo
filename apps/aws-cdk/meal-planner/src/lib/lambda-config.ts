import { Duration } from 'aws-cdk-lib';
import {
  LambdaDefinition,
  LambdaDefinitionProps,
  LambdaFunctionProps,
} from '../lambda-layer/types';
import { NodejsFunctionProps } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import * as path from 'path';
import { UserPool } from 'aws-cdk-lib/aws-cognito';
import { HttpMethod } from 'aws-cdk-lib/aws-stepfunctions-tasks';

// Returns lambda definitions with custom env
export const getLambdaDefinitions = (
  id: string,
  userPool?: UserPool
): LambdaDefinition[] => {
  const environment = {
    DDB_TABLE: `${id}`,
  };
  const lambdaDefinitions: LambdaDefinition[] = [
    // {
    //   name: 'get-users',
    //   environment,
    //   api: {
    //     path: '/users',
    //     methods: [HttpMethod.GET],
    //   },
    // },
    {
      name: 'add-user',
      environment,
      api: {
        path: '/users',
        methods: [HttpMethod.POST],
      },
    },
    // {
    //   name: 'delete-user',
    //   environment,
    //   api: {
    //     path: '/users',
    //     methods: [HttpMethod.DELETE],
    //   },
    // },
    // {
    //   name: 'update-user',
    //   environment,
    //   api: {
    //     path: '/users',
    //     methods: [HttpMethod.PATCH],
    //   },
    // },
    // {
    //   name: 'api-authorizer',
    //   environment: { ...environment, USER_POOL_ID: userPool.userPoolId },
    // },
  ];
  return lambdaDefinitions;
};

// Returns Lambda Function properties with defaults and overwrites
export const getFunctionProps = ({
  id,
  lambdaDefinition,
  lambdaRole,
  lambdaLayer,
}: LambdaFunctionProps): NodejsFunctionProps => {
  const functionProps: NodejsFunctionProps = {
    functionName: `${id}-${lambdaDefinition.name}`,
    entry: path.join(
      __dirname,
      '..',
      'lambda-handlers',
      `${lambdaDefinition.name}.ts`
    ),
    runtime: Runtime.NODEJS_18_X,
    role: lambdaRole,
    layers: [lambdaLayer],
  };
  return functionProps;
};
