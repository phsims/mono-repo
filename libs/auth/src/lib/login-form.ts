export class LoginForm extends HTMLElement {
  connectedCallback() {
    this.innerHTML = `
    <div>
    <input type="email" />
    <input type="password" />
    <button/>
    </div>`;
  }
}

customElements.define('login-form', LoginForm);
