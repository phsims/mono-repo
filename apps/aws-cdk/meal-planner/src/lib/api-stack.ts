import { Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';

import {
  HttpLambdaAuthorizer,
  HttpLambdaResponseType,
} from '@aws-cdk/aws-apigatewayv2-authorizers-alpha';
import { HttpApi, CorsHttpMethod } from '@aws-cdk/aws-apigatewayv2-alpha';

import { HttpLambdaIntegration } from '@aws-cdk/aws-apigatewayv2-integrations-alpha';

import { getLambdaDefinitions } from './lambda-config';

// Types
import { APIStackProps } from '../lambda-layer/types';

export class APIStack extends Stack {
  constructor(scope: Construct, id: string, props: APIStackProps) {
    super(scope, id, props);

    // Define API Authorizer
    const apiAuthorizer = new HttpLambdaAuthorizer(
      'apiAuthorizer',
      props.lambdaFunctions['api-authorizer'],
      {
        authorizerName: `${id}-http-api-authorizer`,
        responseTypes: [HttpLambdaResponseType.SIMPLE],
      }
    );

    // // Define HTTP API
    const httpApi = new HttpApi(this, 'httpApi', {
      apiName: `${id}-api`,
      description: `HTTP API - ${id}`,
      corsPreflight: {
        allowHeaders: ['Authorization', 'Content-Type'],
        allowMethods: [
          CorsHttpMethod.GET,
          CorsHttpMethod.POST,
          CorsHttpMethod.DELETE,
          CorsHttpMethod.PATCH,
        ],
        allowOrigins: ['*'],
      },
      defaultAuthorizer: apiAuthorizer,
    });

    // // Get Lambda definitions
    const lambdaDefinitions = getLambdaDefinitions(id, props.userPool);

    // // Loop through lambda definitions and create api routes if any
    for (const lambdaDefinition of lambdaDefinitions) {
      if (lambdaDefinition.api) {
        httpApi.addRoutes({
          path: lambdaDefinition.api.path,
          methods: lambdaDefinition.api.methods,
          integration: new HttpLambdaIntegration(
            `${lambdaDefinition.name}-integration`,
            props.lambdaFunctions[lambdaDefinition.name]
          ),
        });
      }
    }
  }
}
