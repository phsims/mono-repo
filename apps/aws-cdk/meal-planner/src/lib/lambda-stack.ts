
import { Stack, RemovalPolicy } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import {
  Role,
  ServicePrincipal,
  ManagedPolicy,
  Policy,
  PolicyStatement,
  Effect,
} from 'aws-cdk-lib/aws-iam';
import { LayerVersion, Code, Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { LogGroup, RetentionDays } from 'aws-cdk-lib/aws-logs';
import * as path from 'path';

import { LambdaStackProps } from '../lambda-layer/types';
import { getLambdaDefinitions, getFunctionProps } from './lambda-config';

export class LambdaStack extends Stack {
  public readonly lambdaFunctions: {
    [key: string]: NodejsFunction;
  } = {};
  constructor(scope: Construct, id: string, props: LambdaStackProps) {
    super(scope, id, props);

    // Lambda Role
    const lambdaRole = new Role(this, 'lambdaRole', {
      roleName: `${id}-lambda-role`,
      description: `Lambda role for ${id}`,
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName('ReadOnlyAccess'),
      ],
    });

    // Attach inline policies to Lambda role
    lambdaRole.attachInlinePolicy(
      new Policy(this, 'lambdaExecutionAccess', {
        policyName: 'lambdaExecutionAccess',
        statements: [
          new PolicyStatement({
            effect: Effect.ALLOW,
            resources: ['*'],
            actions: [
              'logs:CreateLogGroup',
              'logs:CreateLogStream',
              'logs:DescribeLogGroups',
              'logs:DescribeLogStreams',
              'logs:PutLogEvents',
            ],
          }),
        ],
      })
    );

    // Add DDB read write permissions to lambda role
    props.table.grantReadWriteData(lambdaRole);

    // Lambda Layer
    const lambdaLayer = new LayerVersion(this, 'lambdaLayer', {
      code: Code.fromAsset(path.join(__dirname, '..', 'lambda-layer')),
      compatibleRuntimes: [Runtime.NODEJS_18_X],
      description: `Lambda Layer for ${id}`,
    });

    // Get Lambda definitions
    const lambdaDefinitions = getLambdaDefinitions(id, props.userPool);

    // Loop through the definitions and create lambda functions
    for (const lambdaDefinition of lambdaDefinitions) {
      // Get function props based on lambda definition
      let functionProps = getFunctionProps({
        id,
        lambdaDefinition,
        lambdaRole,
        lambdaLayer,
      });

      // Lambda Function
      const lambdaFunction = new NodejsFunction(
        this,
        `${lambdaDefinition.name}-function`,
        functionProps
      );
      this.lambdaFunctions[lambdaDefinition.name] = lambdaFunction;

      // Create corresponding Log Group with one month retention
      new LogGroup(this, `fn-${lambdaDefinition.name}-log-group`, {
        logGroupName: `/aws/lambda/${id}-${lambdaDefinition.name}`,
        retention: RetentionDays.ONE_MONTH,
        removalPolicy: RemovalPolicy.DESTROY,
      });
    }
  }
}
