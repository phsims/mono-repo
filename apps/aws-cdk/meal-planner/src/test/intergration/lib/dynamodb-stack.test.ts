import { Stack, App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { DynamoDbStack } from '../../../lib/dynamodb-stack';

test('DynamoDb Stack', () => {
  const app = new App();

  const stack = new DynamoDbStack(app, 'TestStack');
  const template = Template.fromStack(stack as Stack);

  expect(template).toMatchSnapshot('TestStack');
});
