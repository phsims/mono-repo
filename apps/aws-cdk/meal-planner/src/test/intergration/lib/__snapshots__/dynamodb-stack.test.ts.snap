// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamoDb Stack: TestStack 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "TestStacktable8097466F": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "PK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "SK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "GSI1PK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "GSI1SK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "GSI2PK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "GSI2SK",
            "AttributeType": "N",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "GlobalSecondaryIndexes": [
          {
            "IndexName": "GSI1",
            "KeySchema": [
              {
                "AttributeName": "GSI1PK",
                "KeyType": "HASH",
              },
              {
                "AttributeName": "GSI1SK",
                "KeyType": "RANGE",
              },
            ],
            "Projection": {
              "ProjectionType": "ALL",
            },
          },
          {
            "IndexName": "GSI2",
            "KeySchema": [
              {
                "AttributeName": "GSI2PK",
                "KeyType": "HASH",
              },
              {
                "AttributeName": "GSI2SK",
                "KeyType": "RANGE",
              },
            ],
            "Projection": {
              "ProjectionType": "ALL",
            },
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "PK",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "SK",
            "KeyType": "RANGE",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
